README.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> Redis components for Node.js

This module is a part of the [Pip.Services](http://pipservices.org) polyglot microservices toolkit.

The module for working with a Redis database of type key-value contains components for working with cache and locks: RedisCache and RedisLock.

The module contains the following packages:

- **Build** - factory default
- **Cache** - Redis Cache Components
- **Lock** - components of working with locks in Redis

<a name="links"></a> Quick links:

* [Configuration](https://www.pipservices.org/recipies/configuration)
* [API Reference](https://pip-services3-node.github.io/pip-services3-redis-node/globals.html)
* [Change Log](CHANGELOG.md)
* [Get Help](https://www.pipservices.org/community/help)
* [Contribute](https://www.pipservices.org/community/contribute)

## Use

Install the NPM package as
```bash
npm install pip-services-redis-node --save
```

## Develop

For development you shall install the following prerequisites:
* Node.js 8+
* Visual Studio Code or another IDE of your choice
* Docker
* Typescript

Install dependencies:
```bash
npm install
```

Compile the code:
```bash
tsc
```

Run automated tests:
```bash
npm test
```

Generate API documentation:
```bash
./docgen.ps1
```

Before committing changes run dockerized build and test as:
```bash
./build.ps1
./test.ps1
./clear.ps1
```

## Contacts

The library is created and maintained by **Sergey Seroukhov**.


CHANGELOG.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> Redis components for Node.js Changelog

## <a name="3.0.5"></a> 3.0.5 (2021-10-28) 

### Bug fixing
* Fix working with objects, added tests
## <a name="3.0.0"></a> 3.0.0 (2018-11-03) 

### Breaking Changes
* Moved to a separate package

## <a name="1.2.0"></a> 1.2.0 (2018-08-10) 

### Features
* Added RedisCache
* Added RedisLock



test/lock/RedisLock.test.ts
import { ConfigParams } from 'pip-services3-commons-node';

import { RedisLock } from '../../src/lock/RedisLock';
import { LockFixture } from '../fixtures/LockFixture';

suite('RedisLock', ()=> {
    var _lock: RedisLock;
    var _fixture: LockFixture;

    setup((done) => {
        let host = process.env['REDIS_SERVICE_HOST'] || 'localhost';
        let port = process.env['REDIS_SERVICE_PORT'] || 6379;

        _lock = new RedisLock();

        let config = ConfigParams.fromTuples(
            'connection.host', host,
            'connection.port', port
        );
        _lock.configure(config);

        _fixture = new LockFixture(_lock);

        _lock.open(null, done);
    });

    teardown((done) => {
        _lock.close(null, done);
    });

    test('Try Acquire Lock', (done) => {
        _fixture.testTryAcquireLock(done);
    });

    test('Acquire Lock', (done) => {
        _fixture.testAcquireLock(done);
    });

    test('Release Lock', (done) => {
        _fixture.testReleaseLock(done);
    });

});


test/cache/RedisCache.test.ts
let process = require('process');
let assert = require('chai').assert;
let async = require('async');

import { Descriptor } from 'pip-services3-commons-node';
import { ConfigParams } from 'pip-services3-commons-node';

import { RedisCache } from '../../src/cache/RedisCache';
import { CacheFixture } from '../fixtures/CacheFixture';

suite('RedisCache', ()=> {
    let _cache: RedisCache;
    let _fixture: CacheFixture;

    setup((done) => {
        let host = process.env['REDIS_SERVICE_HOST'] || 'localhost';
        let port = process.env['REDIS_SERVICE_PORT'] || 6379;

        _cache = new RedisCache();

        let config = ConfigParams.fromTuples(
            'connection.host', host,
            'connection.port', port
        );
        _cache.configure(config);

        _fixture = new CacheFixture(_cache);

        _cache.open(null, done);
    });

    teardown((done) => {
        _cache.close(null, done);
    });

    test('Store and Retrieve', (done) => {
        _fixture.testStoreAndRetrieve(done);
    });    

    test('Retrieve Expired', (done) => {
        _fixture.testRetrieveExpired(done);
    });    

    test('Remove', (done) => {
        _fixture.testRemove(done);
    });    
    
});


test/fixtures/CacheFixture.ts
let assert = require('chai').assert;
let async = require('async');

import { ICache } from 'pip-services3-components-node';

let KEY1: string = "key1";
let KEY2: string = "key2";
let KEY3: string = "key3";
let KEY4: string = "key4";
let KEY5: string = "key5";
let KEY6: string = "key6";

let VALUE1: string = "value1";
let VALUE2 = { val: "value2" };
let VALUE3 = new Date();
let VALUE4 = [1, 2, 3, 4];
let VALUE5 = 12345;
let VALUE6 = null;

export class CacheFixture {
    private _cache: ICache = null;

    public constructor(cache: ICache) {
        this._cache = cache;
    }

    public testStoreAndRetrieve(done: any): void {
        async.series([
            (callback) => {
                this._cache.store(null, KEY1, VALUE1, 5000, (err) => {
                    assert.isNull(err || null);
                    callback();
                });
            },
            (callback) => {
                this._cache.store(null, KEY2, VALUE2, 5000, (err) => {
                    assert.isNull(err || null);
                    callback();
                });
            },
            (callback) => {
                this._cache.store(null, KEY3, VALUE3, 5000, (err) => {
                    assert.isNull(err || null);
                    callback();
                });
            },
            (callback) => {
                this._cache.store(null, KEY4, VALUE4, 5000, (err) => {
                    assert.isNull(err || null);
                    callback();
                });
            },
            (callback) => {
                this._cache.store(null, KEY5, VALUE5, 5000, (err) => {
                    assert.isNull(err || null);
                    callback();
                });
            },
            (callback) => {
                this._cache.store(null, KEY6, VALUE6, 5000, (err) => {
                    assert.isNull(err || null);
                    callback();
                });
            },
            (callback) => {
                setTimeout(() => {
                    callback();
                }, 500);
            },
            (callback) => {
                this._cache.retrieve(null, KEY1, (err, val) => {
                    assert.isNull(err || null);
                    assert.isNotNull(val);
                    assert.equal(VALUE1, val);

                    callback();
                });
            },
            (callback) => {
                this._cache.retrieve(null, KEY2, (err, val) => {
                    assert.isNull(err || null);
                    assert.isNotNull(val);
                    assert.equal(VALUE2.val, val.val);

                    callback();
                });
            },
            (callback) => {
                this._cache.retrieve(null, KEY3, (err, val) => {
                    assert.isNull(err || null);
                    assert.isNotNull(val);
                    assert.equal(VALUE3.toISOString(), val);

                    callback();
                });
            },
            (callback) => {
                this._cache.retrieve(null, KEY4, (err, val) => {
                    assert.isNull(err || null);
                    assert.isNotNull(val);
                    assert.lengthOf(val, 4)
                    assert.equal(VALUE4[0], val[0]);

                    callback();
                });
            },
            (callback) => {
                this._cache.retrieve(null, KEY5, (err, val) => {
                    assert.isNull(err || null);
                    assert.isNotNull(val);
                    assert.equal(VALUE5, val);

                    callback();
                });
            },
            (callback) => {
                this._cache.retrieve(null, KEY6, (err, val) => {
                    assert.isNull(err || null);
                    assert.isNull(val);

                    callback();
                });
            }
        ], done);
    }

    public testRetrieveExpired(done: any): void {
        async.series([
            (callback) => {
                this._cache.store(null, KEY1, VALUE1, 1000, (err) => {
                    assert.isNull(err || null);
                    callback();
                });
            },
            (callback) => {
                setTimeout(() => {
                    callback();
                }, 1500);
            },
            (callback) => {
                this._cache.retrieve(null, KEY1, (err, val) => {
                    assert.isNull(err || null);
                    assert.isNull(val || null);

                    callback();
                });
            }
        ], done);
    }

    public testRemove(done: any): void {
        async.series([
            (callback) => {
                this._cache.store(null, KEY1, VALUE1, 1000, (err) => {
                    assert.isNull(err || null);
                    callback();
                });
            },
            (callback) => {
                this._cache.remove(null, KEY1, (err) => {
                    assert.isNull(err || null);
                    callback();
                });
            },
            (callback) => {
                this._cache.retrieve(null, KEY1, (err, val) => {
                    assert.isNull(err || null);
                    assert.isNull(val || null);

                    callback();
                });
            }
        ], done);
    }

}


src/lock/RedisLock.ts
/** @module lock */
/** @hidden */
const _ = require('lodash');
/** @hidden */
const async = require('async');

import { ConfigParams } from 'pip-services3-commons-node';
import { IConfigurable } from 'pip-services3-commons-node';
import { IReferences } from 'pip-services3-commons-node';
import { IReferenceable } from 'pip-services3-commons-node';
import { IOpenable } from 'pip-services3-commons-node';
import { IdGenerator } from 'pip-services3-commons-node';
import { InvalidStateException } from 'pip-services3-commons-node';
import { ConfigException } from 'pip-services3-commons-node';
import { ConnectionParams } from 'pip-services3-components-node';
import { ConnectionResolver } from 'pip-services3-components-node';
import { CredentialParams } from 'pip-services3-components-node';
import { CredentialResolver } from 'pip-services3-components-node';
import { Lock } from 'pip-services3-components-node';

/**
 * Distributed lock that is implemented based on Redis in-memory database.
 * 
 * ### Configuration parameters ###
 * 
 * - connection(s):           
 *   - discovery_key:         (optional) a key to retrieve the connection from [[https://pip-services3-node.github.io/pip-services3-components-node/interfaces/connect.idiscovery.html IDiscovery]]
 *   - host:                  host name or IP address
 *   - port:                  port number
 *   - uri:                   resource URI or connection string with all parameters in it
 * - credential(s):
 *   - store_key:             key to retrieve parameters from credential store
 *   - username:              user name (currently is not used)
 *   - password:              user password
 * - options:
 *   - retry_timeout:         timeout in milliseconds to retry lock acquisition. (Default: 100)
 *   - retries:               number of retries (default: 3)
 *  
 * ### References ###
 * 
 * - <code>\*:discovery:\*:\*:1.0</code>        (optional) [[https://pip-services3-node.github.io/pip-services3-components-node/interfaces/connect.idiscovery.html IDiscovery]] services to resolve connection
 * - <code>\*:credential-store:\*:\*:1.0</code> (optional) Credential stores to resolve credential
 *
 * ### Example ###
 * 
 *     let lock = new RedisRedis();
 *     lock.configure(ConfigParams.fromTuples(
 *       "host", "localhost",
 *       "port", 6379
 *     ));
 * 
 *     lock.open("123", (err) => {
 *       ...
 *     });
 * 
 *     lock.acquire("123", "key1", (err) => {
 *          if (err == null) {
 *              try {
 *                // Processing...
 *              } finally {
 *                 lock.releaseLock("123", "key1", (err) => {
 *                     // Continue...
 *                 });
 *              }
 *          }
 *     });
 */
export class RedisLock extends Lock implements IConfigurable, IReferenceable, IOpenable {
    private _connectionResolver: ConnectionResolver = new ConnectionResolver();
    private _credentialResolver: CredentialResolver = new CredentialResolver();
    
    private _lock: string = IdGenerator.nextLong();
    private _timeout: number = 30000;
    private _retries: number = 3;

    private _client: any = null;

    /**
     * Configures component by passing configuration parameters.
     * 
     * @param config    configuration parameters to be set.
     */
    public configure(config: ConfigParams): void {
        this._connectionResolver.configure(config);
        this._credentialResolver.configure(config);

        this._timeout = config.getAsIntegerWithDefault('options.timeout', this._timeout);
        this._retries = config.getAsIntegerWithDefault('options.retries', this._retries);
    }

    /**
	 * Sets references to dependent components.
	 * 
	 * @param references 	references to locate the component dependencies. 
     */
    public setReferences(references: IReferences): void {
        this._connectionResolver.setReferences(references);
        this._credentialResolver.setReferences(references);
    }

    /**
	 * Checks if the component is opened.
	 * 
	 * @returns true if the component has been opened and false otherwise.
     */
    public isOpen(): boolean {
        return this._client;
    }

    /**
	 * Opens the component.
	 * 
	 * @param correlationId 	(optional) transaction id to trace execution through call chain.
     * @param callback 			callback function that receives error or null no errors occured.
     */
    public open(correlationId: string, callback: (err: any) => void): void {
        let connection: ConnectionParams;
        let credential: CredentialParams;

        async.series([
            (callback) => {
                this._connectionResolver.resolve(correlationId, (err, result) => {
                    connection = result;
                    if (err == null && connection == null)
                        err = new ConfigException(correlationId, 'NO_CONNECTION', 'Connection is not configured');
                    callback(err);
                });
            },
            (callback) => {
                this._credentialResolver.lookup(correlationId, (err, result) => {
                    credential = result;
                    callback(err);
                });
            },
            (callback) => {
                let options: any = {
                    // connect_timeout: this._timeout,
                    // max_attempts: this._retries,
                    retry_strategy: (options) => { return this.retryStrategy(options); }
                };
                
                if (connection.getUri() != null) {
                    options.url = connection.getUri();
                } else {                    
                    options.host = connection.getHost() || 'localhost';
                    options.port = connection.getPort() || 6379;
                }

                if (credential != null) {
                    options.password = credential.getPassword();
                }
    
                let redis = require('redis');
                this._client = redis.createClient(options);
    
                if (callback) callback(null);    
            }
        ], callback);
    }

    /**
	 * Closes component and frees used resources.
	 * 
	 * @param correlationId 	(optional) transaction id to trace execution through call chain.
     * @param callback 			callback function that receives error or null no errors occured.
     */
    public close(correlationId: string, callback: (err: any) => void): void {
        if (this._client != null) {
            this._client.quit(((err) => {
                this._client = null;    
                if (callback) callback(err);
            }));
        } else {
            if (callback) callback(null);
        }
    }

    private checkOpened(correlationId: string, callback: any): boolean {
        if (!this.isOpen()) {
            let err = new InvalidStateException(correlationId, 'NOT_OPENED', 'Connection is not opened');
            callback(err, null);
            return false;
        }
        
        return true;
    }
    
    private retryStrategy(options: any): any {
        if (options.error && options.error.code === 'ECONNREFUSED') {
            // End reconnecting on a specific error and flush all commands with
            // a individual error
            return new Error('The server refused the connection');
        }
        if (options.total_retry_time > this._timeout) {
            // End reconnecting after a specific timeout and flush all commands
            // with a individual error
            return new Error('Retry time exhausted');
        }
        if (options.attempt > this._retries) {
            // End reconnecting with built in error
            return undefined;
        }
        // reconnect after
        return Math.min(options.attempt * 100, 3000);
    }

    /**
     * Makes a single attempt to acquire a lock by its key.
     * It returns immediately a positive or negative result.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param key               a unique lock key to acquire.
     * @param ttl               a lock timeout (time to live) in milliseconds.
     * @param callback          callback function that receives a lock result or error.
     */
    public tryAcquireLock(correlationId: string, key: string, ttl: number,
        callback: (err: any, result: boolean) => void): void {
        if (!this.checkOpened(correlationId, callback)) return;

        this._client.set(key, this._lock, 'NX', 'PX', ttl, (err, result) => {
            callback(err, result == "OK");
        });
    }

    /**
     * Releases prevously acquired lock by its key.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param key               a unique lock key to release.
     * @param callback          callback function that receives error or null for success.
     */
    public releaseLock(correlationId: string, key: string,
        callback?: (err: any) => void): void {
        if (!this.checkOpened(correlationId, callback)) return;

        // Start transaction on key
        this._client.watch(key, (err) => {
            if (err) {
                if (callback) callback(err);
                return;
            }

            // Read and check if lock is the same
            this._client.get(key, (err, result) => {
                if (err) {
                    if (callback) callback(err);
                    return;                    
                }

                // Remove the lock if it matches
                if (result == this._lock) {
                    this._client.multi()
                        .del(key)
                        .exec(callback);
                }
                // Cancel transaction if it doesn't match
                else {
                    this._client.unwatch(callback);
                }
            })
        });
    }    
}

src/cache/RedisCache.ts
/** @module cache */
/** @hidden */
const _ = require('lodash');
/** @hidden */
const async = require('async');

import { ConfigParams } from 'pip-services3-commons-node';
import { IConfigurable } from 'pip-services3-commons-node';
import { IReferences } from 'pip-services3-commons-node';
import { IReferenceable } from 'pip-services3-commons-node';
import { IOpenable } from 'pip-services3-commons-node';
import { InvalidStateException } from 'pip-services3-commons-node';
import { ConfigException } from 'pip-services3-commons-node';
import { ConnectionParams } from 'pip-services3-components-node';
import { ConnectionResolver } from 'pip-services3-components-node';
import { CredentialParams } from 'pip-services3-components-node';
import { CredentialResolver } from 'pip-services3-components-node';
import { ICache } from 'pip-services3-components-node';

/**
 * Distributed cache that stores values in Redis in-memory database.
 * 
 * ### Configuration parameters ###
 * 
 * - connection(s):           
 *   - discovery_key:         (optional) a key to retrieve the connection from [[https://pip-services3-node.github.io/pip-services3-components-node/interfaces/connect.idiscovery.html IDiscovery]]
 *   - host:                  host name or IP address
 *   - port:                  port number
 *   - uri:                   resource URI or connection string with all parameters in it
 * - credential(s):
 *   - store_key:             key to retrieve parameters from credential store
 *   - username:              user name (currently is not used)
 *   - password:              user password
 * - options:
 *   - retries:               number of retries (default: 3)
 *   - timeout:               default caching timeout in milliseconds (default: 1 minute)
 *   - max_size:              maximum number of values stored in this cache (default: 1000)        
 *  
 * ### References ###
 * 
 * - <code>\*:discovery:\*:\*:1.0</code>        (optional) [[https://pip-services3-node.github.io/pip-services3-components-node/interfaces/connect.idiscovery.html IDiscovery]] services to resolve connection
 * - <code>\*:credential-store:\*:\*:1.0</code> (optional) Credential stores to resolve credential
 *
 * ### Example ###
 * 
 *     let cache = new RedisCache();
 *     cache.configure(ConfigParams.fromTuples(
 *       "host", "localhost",
 *       "port", 6379
 *     ));
 * 
 *     cache.open("123", (err) => {
 *       ...
 *     });
 * 
 *     cache.store("123", "key1", "ABC", (err) => {
 *          cache.store("123", "key1", (err, value) => {
 *              // Result: "ABC"
 *          });
 *     });
 */
export class RedisCache implements ICache, IConfigurable, IReferenceable, IOpenable {
    private _connectionResolver: ConnectionResolver = new ConnectionResolver();
    private _credentialResolver: CredentialResolver = new CredentialResolver();

    private _timeout: number = 30000;
    private _retries: number = 3;

    private _client: any = null;

    /**
     * Creates a new instance of this cache.
     */
    public constructor() { }

    /**
     * Configures component by passing configuration parameters.
     * 
     * @param config    configuration parameters to be set.
     */
    public configure(config: ConfigParams): void {
        this._connectionResolver.configure(config);
        this._credentialResolver.configure(config);

        this._timeout = config.getAsIntegerWithDefault('options.timeout', this._timeout);
        this._retries = config.getAsIntegerWithDefault('options.retries', this._retries);
    }

    /**
     * Sets references to dependent components.
     * 
     * @param references 	references to locate the component dependencies. 
     */
    public setReferences(references: IReferences): void {
        this._connectionResolver.setReferences(references);
        this._credentialResolver.setReferences(references);
    }

    /**
     * Checks if the component is opened.
     * 
     * @returns true if the component has been opened and false otherwise.
     */
    public isOpen(): boolean {
        return this._client;
    }

    /**
     * Opens the component.
     * 
     * @param correlationId 	(optional) transaction id to trace execution through call chain.
     * @param callback 			callback function that receives error or null no errors occured.
     */
    public open(correlationId: string, callback: (err: any) => void): void {
        let connection: ConnectionParams;
        let credential: CredentialParams;

        async.series([
            (callback) => {
                this._connectionResolver.resolve(correlationId, (err, result) => {
                    connection = result;
                    if (err == null && connection == null)
                        err = new ConfigException(correlationId, 'NO_CONNECTION', 'Connection is not configured');
                    callback(err);
                });
            },
            (callback) => {
                this._credentialResolver.lookup(correlationId, (err, result) => {
                    credential = result;
                    callback(err);
                });
            },
            (callback) => {
                let options: any = {
                    // connect_timeout: this._timeout,
                    // max_attempts: this._retries,
                    retry_strategy: (options) => { return this.retryStrategy(options); }
                };

                if (connection.getUri() != null) {
                    options.url = connection.getUri();
                } else {
                    options.host = connection.getHost() || 'localhost';
                    options.port = connection.getPort() || 6379;
                }

                if (credential != null) {
                    options.password = credential.getPassword();
                }

                let redis = require('redis');
                this._client = redis.createClient(options);

                if (callback) callback(null);
            }
        ], callback);
    }

    /**
     * Closes component and frees used resources.
     * 
     * @param correlationId 	(optional) transaction id to trace execution through call chain.
     * @param callback 			callback function that receives error or null no errors occured.
     */
    public close(correlationId: string, callback: (err: any) => void): void {
        if (this._client != null) {
            this._client.quit(((err) => {
                this._client = null;
                if (callback) callback(err);
            }));
        } else {
            if (callback) callback(null);
        }
    }

    private checkOpened(correlationId: string, callback: any): boolean {
        if (!this.isOpen()) {
            let err = new InvalidStateException(correlationId, 'NOT_OPENED', 'Connection is not opened');
            callback(err, null);
            return false;
        }

        return true;
    }

    private retryStrategy(options: any): any {
        if (options.error && options.error.code === 'ECONNREFUSED') {
            // End reconnecting on a specific error and flush all commands with
            // a individual error
            return new Error('The server refused the connection');
        }
        if (options.total_retry_time > this._timeout) {
            // End reconnecting after a specific timeout and flush all commands
            // with a individual error
            return new Error('Retry time exhausted');
        }
        if (options.attempt > this._retries) {
            // End reconnecting with built in error
            return undefined;
        }
        // reconnect after
        return Math.min(options.attempt * 100, 3000);
    }

    /**
     * Retrieves cached value from the cache using its key.
     * If value is missing in the cache or expired it returns null.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param key               a unique value key.
     * @param callback          callback function that receives cached value or error.
     */
    public retrieve(correlationId: string, key: string,
        callback: (err: any, value: any) => void): void {
        if (!this.checkOpened(correlationId, callback)) return;

        this._client.get(key, (err, item) => {
            if (err) {
                callback(err, null);
                return;
            }
            callback(err, item ? JSON.parse(item) : item);
        });
    }

    /**
     * Stores value in the cache with expiration time.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param key               a unique value key.
     * @param value             a value to store.
     * @param timeout           expiration timeout in milliseconds.
     * @param callback          (optional) callback function that receives an error or null for success
     */
    public store(correlationId: string, key: string, value: any, timeout: number,
        callback: (err: any) => void): void {
        if (!this.checkOpened(correlationId, callback)) return;

        this._client.set(key, JSON.stringify(value), 'PX', timeout, callback);
    }

    /**
     * Removes a value from the cache by its key.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param key               a unique value key.
     * @param callback          (optional) callback function that receives an error or null for success
     */
    public remove(correlationId: string, key: string,
        callback: (err: any) => void) {
        if (!this.checkOpened(correlationId, callback)) return;

        this._client.del(key, callback);
    }

}